azurerm/resource_arm_iothub_resource_provider.go

```diff
@@ -0,0 +1,261 @@
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"shared_access_policy": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"key_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"primary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"secondary_key": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"permissions": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+}
+
+func resourceArmIotHubCreateAndUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	rg := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+
+	res, err := iothubClient.CheckNameAvailability(ctx, devices.OperationInputs{
+		Name: &name,
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if !*res.NameAvailable {
+		return errors.New(string(res.Reason))
+	}
+
+	location := d.Get("location").(string)
+
+	subscriptionID := armClient.subscriptionId
+	skuInfo := expandAzureRmIotHubSku(d)
+
+	desc := devices.IotHubDescription{
+		Resourcegroup:  &rg,
+		Name:           &name,
+		Location:       &location,
+		Subscriptionid: &subscriptionID,
+		Sku:            &skuInfo,
+	}
+
+	if tagsI, ok := d.GetOk("tags"); ok {
+		tags := tagsI.(map[string]interface{})
+		desc.Tags = *expandTags(tags)
+	}
+
+	match := ""
+
+	if etagI, ok := d.GetOk("etag"); ok {
+		etag := etagI.(string)
+		desc.Etag = &etag
+		match = etag
+	}
+
+	future, err := iothubClient.CreateOrUpdate(ctx, rg, name, desc, match)
+	if err != nil {
+		return err
+	}
+
+	err = future.WaitForCompletion(ctx, iothubClient.Client)
+	if err != nil {
+		return fmt.Errorf("Error creating or updating IotHub %q (Resource Group %q): %+v", name, rg, err)
+	}
+
+	desc, err = iothubClient.Get(ctx, rg, name)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(*desc.ID)
+	return resourceArmIotHubRead(d, meta)
+
+}
+
+func expandAzureRmIotHubSku(d *schema.ResourceData) devices.IotHubSkuInfo {
+	skuList := d.Get("sku").([]interface{})
+	skuMap := skuList[0].(map[string]interface{})
+	cap := int64(skuMap["capacity"].(int))
+
+	return devices.IotHubSkuInfo{
+		Name:     devices.IotHubSku(skuMap["name"].(string)),
+		Tier:     devices.IotHubSkuTier(skuMap["tier"].(string)),
+		Capacity: &cap,
+	}
+
+}
+
+func resourceArmIotHubRead(d *schema.ResourceData, meta interface{}) error {
+	armClient := meta.(*ArmClient)
+	iothubClient := armClient.iothubResourceClient
+	ctx := armClient.StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	iothubName := id.Path["IotHubs"]
+	desc, err := iothubClient.Get(ctx, id.ResourceGroup, iothubName)
+
+	if err != nil {
+		return err
+	}
+
+	properties := desc.Properties
```

Comment:
if the Swagger is invalid / the API response varies (such as when reading an old resource) - this isn't guaranteed to be returned, unfortunately. as such could we check for it having a value:

```
if props := desc.Properties; props != nil {
  # ..
}
```